ocaml/quicktest/quicktest_vdi_copy_data_integrity.ml

```diff
@@ -0,0 +1,181 @@
+    raise e
+
+let finally f cleanup =
+  let result =
+    try
+      f ()
+    with e -> begin
+        cleanup ();
+        raise e
+      end
+  in
+  cleanup ();
+  result
+
+let get_domain_zero rpc session_id =
+  Xapi_inventory.inventory_filename := "/etc/xensource-inventory";
+  let uuid = Xapi_inventory.lookup Xapi_inventory._control_domain_uuid in
+  Client.Client.VM.get_by_uuid ~rpc ~session_id ~uuid
+
+let with_attached_vdi rpc session_id vdi mode f =
+  let dom0 = get_domain_zero rpc session_id in
+  let vbd =
+    Client.Client.VBD.create ~rpc ~session_id
+      ~vM:dom0
+      ~empty:false
+      ~vDI:vdi
+      ~userdevice:"autodetect"
+      ~bootable:false
+      ~mode
+      ~_type:`Disk
+      ~unpluggable:true
+      ~qos_algorithm_type:""
+      ~qos_algorithm_params:[]
+      ~other_config:[]
+  in
+  finally
+    (fun () ->
+       Client.Client.VBD.plug ~rpc ~session_id ~self:vbd;
+       finally
+         (fun () -> f ("/dev/" ^ (Client.Client.VBD.get_device ~rpc ~session_id ~self:vbd)))
+         (fun () ->
+            Client.Client.VBD.unplug ~rpc ~session_id ~self:vbd;
+         )
+    )
+    (fun () -> Client.Client.VBD.destroy ~rpc ~session_id ~self:vbd)
+
+let with_open_vdi rpc session_id vdi mode f =
+  with_attached_vdi rpc session_id vdi mode
+    (fun path ->
+       let mode' = match mode with
+         | `RO -> [ Unix.O_RDONLY ]
+         | `RW -> [ Unix.O_RDWR ] in
+       let fd = Unix.openfile path mode' 0 in
+       finally
+         (fun () -> f fd)
+         (fun () -> Unix.close fd)
+    )
+
+(* ---------------- *
+   Integrity tests
+ * ---------------- *)
+
+let write_random_data rpc session_id vdi =
+  let size = Client.Client.VDI.get_virtual_size ~rpc ~session_id ~self:vdi in
+  with_open_vdi rpc session_id vdi `RW
+    (fun fd ->
+
+       let max_writes = 100 in
+       let max_write_length = 2048 in
+
+       let write_random_char_block offset length =
+         let buf =
+           let c = Random.int 256 |> Char.chr in
+           Bytes.make length c
+         in
+         assert (Unix.LargeFile.lseek fd offset Unix.SEEK_SET = offset);
+         Unix.write fd buf 0 length |> ignore
+       in
+
+       let write_random_extent () =
+         let offset = Random.int64 size in
+         let to_end = Int64.sub size offset in
+         let max_extent_length =
+           min (Int64.of_int max_write_length) to_end |> Int64.to_int
+         in
+         let length = Random.int max_extent_length in
+
+         write_random_char_block offset length
+       in
+
+       let num_extents = Random.int max_writes in
+       for i = 1 to num_extents do
+         write_random_extent ()
+       done
+    )
+
+let fill rpc session_id vdi =
+  let size =
+    Client.Client.VDI.get_virtual_size ~rpc ~session_id ~self:vdi
+    |> Int64.to_int
+  in
+  with_open_vdi rpc session_id vdi `RW
+    (fun fd ->
+       let buf = Bytes.make size 'a' in
+       Unix.write fd buf 0 size
+    )
```

Comment:
Probably I should fill the VDI with more random data, instead of using the same char everywhere. I just couldn't find a library that gives me a random `bytes` of a given length

Reply:
I know only https://github.com/xavierleroy/cryptokit/blob/master/src/cryptokit.mli#L241

Reply:
Or we could copy from `/dev/urandom`?

Reply:
If you look at the various implementations, that's what one of the options in cryptokit does :D
I think we can even use the Random module and fill the buffer.

Reply:
Generating random numbers with a PRNG can be faster than `/dev/urandom` (especially if you just use something fast like a shift register, worse quality randomness but fast), however using `String.init` with `Random.int` is quite slow:
```
String.init Random: 0.225s
/dev/urandom: 0.045s
```

```
let time msg f =
  let t0 = Unix.gettimeofday () in
  ignore(f ());
  let t1 = Unix.gettimeofday () in
  Printf.printf "%s: %.3fs\n" msg (t1 -. t0)

let () =
  Random.self_init ();
  let n = 10_000_000 in

  time "String.init Random" (fun () ->
      String.init n (fun _ -> Char.chr (Random.int 255)));

  time "/dev/urandom" (fun () ->
      let f = open_in "/dev/urandom" in
      let r = really_input_string f n in
      close_in f;
      r)
```

A very fast and simple PRNG with reasonable quality is 2CMRES from here:
http://www.drdobbs.com/tools/fast-high-quality-parallel-random-number/231000484?pgno=2
This one should also be quite fast: http://www.pcg-random.org/
As long as you don't need it for cryptographic purposes either of these algorithms should work well.

Reply:
Thanks for the investigation, I think I'll go for `/dev/urandom` for now. There is also [mirage-random](https://github.com/mirage/mirage-random) in xs-opam, but it uses `Random.int`.